staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go

```diff
@@ -77,6 +79,7 @@ type SelectionPredicate struct {
 	IndexFields []string
 	Limit       int64
 	Continue    string
+	Timeout     time.Duration
```

Comment:
@jpbetz @xiang90 - this is probably the potentially most controversial thing in this PR, as this is changing the storage interface.
I would start with agreeing on that before I will review the whole PR.

My concerns about it are:
(a) if we're going to use that only for watch requests, it should be reflected in name (WatchTimeout?)
(b) if we're going to use it for list requests too, then I would expect some comment for what purpose this is used

I would say that (b) sounds better, as we would also be able to e.g. timeout long list requests and stuff like that.

@jpbetz @xiang90 - I would like to get your buy-in on this interface change.

Reply:
I agree that we should be able to cancel list requests as well. Could we instead use a `context.Context` and support timeouts and cancellation?

Reply:
I believe the main reason to add timeout here is to propagate the actual timeout value down to the cacher/storage layer. A func based context cannot provide this information. 

@wojtek-t @hormes I am not exactly sure how do we propagate the timeout value of list today, or maybe we do not. If that is the case, and we plan to do list timeout, I agree this value should work both for watch and list.

Reply:
> add timeout here is to propagate the actual timeout value down to the cacher/storage layer

It is indeed the case, use for calculating the bookmark period. The timeout isn't for list or watch either, actually we don't that in cacher layer now. 

If we plan to do list/watch timeout in cacher layer, we have to change the store interface. Maybe we can add timeout to context.Context now, and do the list/watch timeout in another issue?

Reply:
So if this is not intended to be used as timeout (at least initially), we should call this field Timeout imho - it would be pretty misleading.

> @wojtek-t @hormes I am not exactly sure how do we propagate the timeout value of list today, or maybe we do not. If that is the case, and we plan to do list timeout, I agree this value should work both for watch and list.

We're not propagating it today at all (but we should be). The question is whether we want to do that together now, or maybe, we want to introduce some special-purpose field for that. I'm leaning towards saying that (given that timeout would be super useful), we may try to join those two things.

Reply:
There is a ```context.Context```  in both watch and list methods, but does not support timeouts or cancellation now.  The watch method returns immediately regardless of this timeout, which means we should use a ```context.Context``` here.

I agree that we should be able to timeout support, but if we consider it as another issue, rename it to ```WatchTimeout``` is better here?

Reply:
yes. just name it as timeout and comment it intends to work for both watch and list, but just watch part is implemented.

Reply:
What I started wondering about is whether this should be part of "SelectionPredicate"? Technically it has nothing to do with selection.
So maybe a better option would be to explicitly add "Timeout" parameter to both List() and Watch() methods? [Then we will have clear split and don't need to add to List now]

@xiang90 @jpbetz - thoughts?

Reply:
I like the idea to add timeout parameter to Watch() method, it does not belong to SelectionPredicate. 

> func (c *Cacher) Watch(ctx context.Context, key string, resourceVersion string, pred storage.SelectionPredicate, timeout time.Duration) (watch.Interface, error) {

But adding a timeout to a nonblocking method sounds wired, we need a comment to explain its purpose.

There is a context in List() method, we don't need a timeout there.
> 	List(ctx context.Context, key string, resourceVersion string, p SelectionPredicate, listObj runtime.Object) error

Reply:
> But adding a timeout to a nonblocking method sounds wired, we need a comment to explain its purpose.

I don't agree it weird - but sure, we need to comment.

@hormes - so can you split this change to a separate PR - I don't think we need to bind those two together, and we can much faster merge the timeout addition

Reply:
> I don't agree it weird - but sure, we need to comment.

We already have context passed in for both list and watch, which can notify us when it is already timed out. But context will not tell us how long it will timeout. For watch, we need to know how long it will time out. That is the difference between the two I guess :). /cc @hormes

Reply:
@wojtek-t separate it to: https://github.com/kubernetes/kubernetes/pull/75835